purchase_delivery_split_date/models/purchase.py

```diff
@@ -1,138 +1,109 @@
+        if picking.move_lines and not self._is_valid_picking(picking):
+            pickings = self.order_id.order_line.move_ids.picking_id
+            picking = fields.first(
+                pickings.filtered(lambda p: self._is_valid_picking(p))
             )
+            if not picking:
+                res = self.order_id._prepare_picking()
+                picking = self.env["stock.picking"].create(res)
+        return super()._prepare_stock_moves(picking)
 
-        for key, lines in date_groups:
-            # If a picking is provided, clone it for each key for modularity
-            if picking:
-                copy_vals = self._first_picking_copy_vals(key, lines)
-                picking = first_picking.copy(copy_vals)
-            po_lines = self.env["purchase.order.line"]
-            for line in list(lines):
-                po_lines += line
-            moves += super(PurchaseOrderLine, po_lines)._create_stock_moves(picking)
-        return moves
+    @api.model
+    def _prepare_picking_copy_vals(self):
+        """The data to be copied to new pickings. This method is designed for
+        extensibility."""
+        return {"move_lines": []}
+
+    def _check_still_valid_picking(self):
+        moves = self.move_ids.filtered(
+            lambda m: m.state not in ("draft", "done", "cancel")
+        )
+        pickings = self.order_id.order_line.move_ids.filtered(
+            lambda m: m.state not in ("draft", "done", "cancel")
+        ).picking_id
+        picking_to_cancel = self.env["stock.picking"]
+        for move in moves:
+            if self._is_valid_picking(move.picking_id):
+                # If the move is the only move of the picking, the picking will
+                # be valid (as the date will be computed as the move date) but
+                # the move could possibly have been merged in another existing
+                # picking
+                if move != move.picking_id.move_lines:
+                    continue
+                picking = fields.first(
+                    pickings.filtered(
+                        lambda p: p != move.picking_id and self._is_valid_picking(p)
+                    )
+                )
+                if not picking:
+                    # No other valid picking
+                    continue
+            else:
+                # Find an other valid picking
+                picking = fields.first(
+                    pickings.filtered(lambda p: self._is_valid_picking(p))
+                )
+                if not picking:
+                    copy_vals = self._prepare_picking_copy_vals()
+                    picking = move.picking_id.copy(copy_vals)
+                    pickings |= picking
+            move._do_unreserve()
+            picking_to_cancel |= move.picking_id
+            move.picking_id = picking
+            move._merge_moves()
+            move._action_assign()
+        picking_to_cancel.filtered(lambda p: not p.move_lines).write(
```

Comment:
I think better call `action_cancel()`

Reply:
No, it cannot. It would do nothing as there are no move lines

Reply:
@ndevas The transfer is not canceled. The moves are reassigned to other transfer. Nothing is canceled! The empty picking is closed in a cancel state